ads-cert-examples/scala/src/main/scala/Main.scala

```diff
@@ -0,0 +1,87 @@
+import java.io.{FileInputStream, InputStreamReader}
+import java.nio.charset.StandardCharsets
+import java.security.spec.X509EncodedKeySpec
+import java.security._
+
+import org.apache.commons.codec.binary.Base64
+
+import org.bouncycastle.asn1.ASN1ObjectIdentifier
+import org.bouncycastle.asn1.x9.ECNamedCurveTable
+import org.bouncycastle.jce.provider.BouncyCastleProvider
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter
+import org.bouncycastle.openssl.{PEMKeyPair, PEMParser}
+
+import scala.util.{Failure, Success, Try}
+
+
+object Main {
+
+  private var privateKey: PrivateKey = _
+  private var publicKey: PublicKey = _
+
+  Security.addProvider(new BouncyCastleProvider())
+
+  def main(args: Array[String]): Unit = {
+    loadKeys(args)
+    val message = Stream(
+      "2824f5d3-a2dc-49ec-8358-dbe92bd01ec2", // source.tid
+      "a:180720yz:1807241326", // ad.pcv
+      "newsite.com", // site.domain
+      "", // app.bundle
+      "", // user.consent
+      "vd", // placement
+      "192.168.1.1", // device.ip
+      "", // device.ipv6
+      "", // device.ifa
+      "", // device.ua
+      "" // ad.video player size?
+    ).map(Option(_).getOrElse("")).toList.mkString(":")
+    for {
+      signature <- sign(Signature.getInstance("SHA256withECDSA"), message)
+    } yield {
+      printf("publisher signature: %s\n", Base64.encodeBase64String(signature))
+      Try {
```

Comment:
Here you're wrapping the call to `.verify` with a `Try {}`, but `.verify` already returns a `Try`.  Removing the outer `Try` would mean that the later `b.getOrElse` would not be necessary.  It should also allow you to pattern match against literals -- `Success(true)` and `Success(false)` rather than have the nested match.

Since both the `.sign` and `.verify` return a `Try` and since `Try` is monadic you can combine them into a single for-comprehension, if you're willing to either move the printing of the signature or use some funky syntax.

```scala
val tryResult = for {
  signature <- sign(Signature.getInstance("SHA256withECDSA"), message)
  result <- verify(Signature.getInstance("SHA256withECDSA"), message, signature)
} yield result

tryResult match {
  case Success(true) => println("yay it worked!")
  case Success(false) => println("it didn't work :(")
  case Failure => throw e // or print a message instead of throw, probably more common
}
```